Implemented Respawn Anchor

b1zeyofficial opened the pull request:
## Introduction
This PR implements Respawn Anchor
this is just a contribution to the community

### Related issues & PRs

## Changes

### API changes

Added class `BlockExplodeEvent` called when a block explodes, after the explosion impact has been calculated
Added class `BlockPreExplodeEvent` called when a block wants to explode, before the explosion impact is calculated
Added class `PlayerRespawnAnchorUseEvent` called when a player interacts with a respawn anchor

### Behavioural changes
When RespawnAnchor explodes, Average TPS may drop by 19.24+-.

## Backwards compatibility

## Follow-up

## Tests
https://www.youtube.com/watch?v=aI3ZItR3FY8

b1zeyofficial committed:
implement respawn anchor and explosion almost like in vanilla

b1zeyofficial committed:
Cleanup

b1zeyofficial committed:
hmhm

b1zeyofficial committed:
first time

b1zeyofficial commented:
I think someone will find a use for this, but for now it's closed.

b1zeyofficial committed:
phpstan, unit...

b1zeyofficial committed:
?

b1zeyofficial committed:
code style

b1zeyofficial committed:
CS

b1zeyofficial committed:
...

b1zeyofficial committed:
dependency

b1zeyofficial committed:
test

b1zeyofficial committed:
no comments

b1zeyofficial commented:
It seems to have been fixed

b1zeyofficial committed:
Fix CS?

b1zeyofficial committed:
CS

b1zeyofficial committed:
Update RespawnAnchor.php

b1zeyofficial commented:
I checked php cs fixer locally but found no errors..

dktapps commented on src/world/Explosion.php:
```diff
@@ -221,4 +299,121 @@ public function explodeB() : bool{
+				if($tMaxY > $radius){
+					break;
+				}
+
+				$current = new Vector3($current->getX(), $current->getY() + $stepY, $current->getZ());
+				$tMaxY += $tDeltaY;
+			}else{
+				if($tMaxZ > $radius){
+					break;
+				}
+
+				$current = new Vector3($current->getX(), $current->getY(), $current->getZ() + $stepZ);
+				$tMaxZ += $tDeltaZ;
+			}
+		}
+
+		return false;
+	}
+
+	private function sign(float $d) : float{
+		if($d > 0){
+			return 1;
+		}
+
+		if($d < 0){
+			return -1;
+		}
+
+		return 0;
+	}
+
+	private function boundary(float $start, float $distance) : float{
+		if($distance === 0.0){
+			return INF;
+		}
+
+		return $distance < 0 ?
+			($start - floor($start)) / -$distance :
+			(1 - ($start - floor($start))) / $distance;
+	}
```
This code is basically identical to `VoxelRayTrace` in `pocketmine\math`.

b1zeyofficial committed:
Update Explosion.php

b1zeyofficial committed:
CS

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

dktapps commented on src/world/Explosion.php:
```diff
@@ -327,7 +327,8 @@ private function getSeenPercent(Vector3 $source, Entity $entity) : float{
 
-					if(!$this->raycastHit($source, $target)){
+					$generator = VoxelRayTrace::betweenPoints($source, $target);
+					if (!$generator->valid()) {
```
That is not how generators work

b1zeyofficial replied:
Can I use next() or current() ? Will it fix the error? In my IDE phpstan complained about the old code calling VoxelRayTrace but valid() fixed it.

b1zeyofficial replied:
> Can I use next() or current() ? Will it fix the error? In my IDE phpstan complained about the old code calling VoxelRayTrace but valid() fixed it.

or just use foreach, i will try that when i get home

dktapps replied:
Yes, use foreach

b1zeyofficial committed:
added next() before valid()

b1zeyofficial commented:
It seems like the generator starts and works.

b1zeyofficial committed:
cs

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+use pocketmine\world\Explosion;
+use pocketmine\world\sound\AnchorChargeSound;
+
+class RespawnAnchor extends Opaque{
```
The class must be final

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+	protected int $charges = 0;
+
+	protected function describeBlockOnlyState(RuntimeDataDescriber $w) : void{
+		$w->boundedIntAuto(0, 4, $this->charges);
```
It would be better to constify min and max charge values

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+		return match ($this->charges){
+			1 => 3,
+			2 => 7,
+			default => ($this->charges > 2 ? 15 : 0),
```
This is incorrect. Respawn anchor has light level 11 with 3 charges. The light calculation formula here is `min(0, 3 + 4 * ($this->charges - 1))`

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($item->getTypeId() === VanillaBlocks::GLOWSTONE()->asItem()->getTypeId() && $this->charges < 4){
```
You can use here `ItemTypeIds::fromBlockTypeId()`

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($item->getTypeId() === VanillaBlocks::GLOWSTONE()->asItem()->getTypeId() && $this->charges < 4){
+			if($player !== null && !$player->isCreative()){
```
This check isn't necessary

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+	public function getCharges() : int{ return $this->charges; }
+
+	public function setCharges(int $charges) : self{
+		$this->charges = $charges;
```
You need to check that the charges amount is in range 0-4, otherwise throw an exception

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+			return false;
+		}
+
+		if($this->charges >= 1){
```
```suggestion
		if($this->charges > 0){

```

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+		return false;
+	}
+
+	public function explode(Player $player) : void{
```
This function should be private

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+
+		if($this->charges >= 1){
+			$this->explode($player);
+			$this->position->getWorld()->setBlock($this->position, VanillaBlocks::AIR(), false);
```
This is not needed. You already set the air block in `explode()`

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+			return true;
+		}
+
+		if($player === null){
```
Why do you limit this? Player can be optional in BlockPreExlodeEvent

b1zeyofficial replied:
PHPStan complains about this, should i remove getPlayer from BlockPreExplodeEvent class?

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+		$ev = new BlockPreExplodeEvent($this, 5, $player);
+		$ev->setIncendiary(true);
+	
+		if($ev->isCancelled()){
```
The event will never be fired because `$ev->call()` is missing

ipad54 commented on src/block/VanillaBlocks.php:
```diff
@@ -1642,6 +1643,9 @@ public function getLightLevel() : int{ return 10;}
 
 		self::register("chain", fn(BID $id) => new Chain($id, "Chain", new Info(BreakInfo::pickaxe(5.0, ToolTier::WOOD))));
+
+		$respawnAnchorBreakInfo = new Info(BreakInfo::pickaxe(50.0, ToolTier::DIAMOND, 1200.0));
```
There is no point to make a break info variable just for one block

ipad54 commented on src/entity/object/EndCrystal.php:
```diff
@@ -130,6 +130,11 @@ public function explode() : void{
+
+			if(!isset($explosion->fireIgnitions)){
+				$explosion->fireIgnitions = [];
+			}
```
I don't understand why do you need this. The properties are initialized by default

ipad54 commented on src/entity/object/PrimedTNT.php:
```diff
@@ -122,6 +122,11 @@ public function explode() : void{
+
+			if(!isset($explosion->fireIgnitions)){
+				$explosion->fireIgnitions = [];
+			}
```
Same here

ipad54 commented on src/world/Explosion.php:
```diff
@@ -78,13 +92,23 @@ public function __construct(
+				$this->doesDamage = false;
+				return true;
+			}
+		}
```
I don't think this has anything to do with the respawn anchor explosion.

ipad54 commented on src/world/Explosion.php:
```diff
@@ -78,13 +92,23 @@ public function __construct(
+		$incendiary = $this->fireChance > 0;
+		if($incendiary && !isset($this->fireIgnitions)){
+			$this->fireIgnitions = [];
+		}
```
The property is always initialized

ipad54 commented on src/world/Explosion.php:
```diff
@@ -119,19 +143,23 @@ public function explodeA() : bool{
 									}
 								}
 							}
+							if($incendiary && mt_rand() / mt_getrandmax() <= $this->fireChance){
```
`mt_rand() / mt_getrandmax()` can be replaced with `Utils::getRandomFloat()`

ipad54 commented on src/world/Explosion.php:
```diff
@@ -191,6 +240,19 @@ public function explodeB() : bool{
+
+					$impact *= $netheriteReduction;
+				}
+
```
This is not related to the respawn anchor

ipad54 commented on src/world/Explosion.php:
```diff
@@ -170,18 +220,17 @@ public function explodeB() : bool{
+				$motion = $entityPos->subtractVector($this->source)->normalize();
 
-				$damage = (int) ((($impact * $impact + $impact) / 2) * 8 * $explosionSize + 1);
+				$damage = $this->doesDamage ? max((int) (((($impact * $impact + $impact) / 2) * 8 * $explosionSize) + 1), 0) : 0;
```
This is not related to the respawn anchor

ipad54 commented on src/world/Explosion.php:
```diff
@@ -209,10 +271,26 @@ public function explodeB() : bool{
-				$this->world->setBlockAt($pos->x, $pos->y, $pos->z, $airBlock);
+
+				$this->world->setBlockAt((int) $pos->x, (int) $pos->y, (int) $pos->z, $airBlock);
+			}
```
Unrelated changes

ipad54 commented on src/world/Explosion.php:
```diff
@@ -119,19 +143,23 @@ public function explodeA() : bool{
 										foreach($_block->getAffectedBlocks() as $_affectedBlock){
 											$_affectedBlockPos = $_affectedBlock->getPosition();
-											$this->affectedBlocks[World::blockHash($_affectedBlockPos->x, $_affectedBlockPos->y, $_affectedBlockPos->z)] = $_affectedBlock;
+											$this->affectedBlocks[World::blockHash((int) $_affectedBlockPos->x, (int) $_affectedBlockPos->y, (int) $_affectedBlockPos->z)] = $_affectedBlock;
```
Unrelated change

ipad54 commented on src/world/Explosion.php:
```diff
@@ -119,19 +143,23 @@ public function explodeA() : bool{
+							$block = $this->world->getBlockAt($vBlockX, $vBlockY, $vBlockZ, true, false);
 							$blastResistance = $blockFactory->blastResistance[$state] ?? 0;
-							if($blastResistance >= 0){
+							if($blastResistance >= 0 && $block->getTypeId() !== VanillaBlocks::AIR()->getTypeId()){
```
You don't need to check for air here. Air block always has negative blast resistance value

ipad54 commented on src/world/Explosion.php:
```diff
@@ -54,6 +64,10 @@ class Explosion{
 	public array $affectedBlocks = [];
 	public float $stepLen = 0.3;
+	private bool $doesDamage = true;
+	public float $fireChance = 0.0;
+	/** @var Block[] */
+	public array $fireIgnitions = [];
```
These properties probably should be private

ipad54 commented on src/world/sound/AnchorChargeSound.php:
```diff
@@ -0,0 +1,35 @@
+use pocketmine\network\mcpe\protocol\LevelSoundEventPacket;
+use pocketmine\network\mcpe\protocol\types\LevelSoundEvent;
+
+class AnchorChargeSound implements Sound{
```
I think `RespawnAnchorChargeSound` would be more clear name

b1zeyofficial replied:
This name looks long but oh well

ipad54 commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -0,0 +1,119 @@
+	public function setIncendiary(bool $incendiary) : void{
+		if(!$incendiary){
+			$this->fireChance = 0;
+		}else{
+			if($this->fireChance <= 0){
```
You can combine this into simple `elseif` block

ipad54 commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -0,0 +1,119 @@
+	 * Establish the probability of fire
+	 */
+	public function setFireChance(float $fireChance) : void{
+		$this->fireChance = $fireChance;
```
You need to do a check here that the fire chance value is between 0-1

ipad54 commented on src/event/entity/EntityExplodeEvent.php:
```diff
@@ -42,14 +42,17 @@ class EntityExplodeEvent extends EntityEvent implements Cancellable{
 		protected array $blocks,
-		protected float $yield
+		protected float $yield,
+		/** @var Block[] $ignitions */
```
This var isn't necessary

ipad54 commented on src/event/entity/EntityExplodeEvent.php:
```diff
@@ -98,4 +101,22 @@ public function setYield(float $yield) : void{
+	 * @param Block[] $ignitions
+	 */
+	public function setIgnitions(array $ignitions) : void{
+		$this->ignitions = $ignitions;
```
You need to validate array contents first

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,99 @@
+use pocketmine\world\sound\AnchorChargeSound;
+
+class RespawnAnchor extends Opaque{
+	protected int $charges = 0;
```
This property must be private

b1zeyofficial replied:
should const be private too?

ipad54 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,137 @@
+	 * @param Block[] $blocks The set of blocks to be affected by fire ignitions
+	 */
+	public function setAffectedBlocks(array $blocks) : void{
+		$this->blocks = $blocks;
```
Array contents must be validated

ipad54 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,137 @@
+	 * @param Block[] $ignitions The set of blocks to set as ignited
+	 */
+	public function setIgnitions(array $ignitions) : void{
+		$this->ignitions = $ignitions;
```
Same here

ipad54 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,137 @@
+	 * @param float $fireChance The fire chance to set
+	 */
+	public function setFireChance(float $fireChance) : void{
+		$this->fireChance = $fireChance;
```
You need to do a check here that the fire chance value is between 0-1

ipad54 commented on src/world/Explosion.php:
```diff
@@ -209,10 +271,26 @@ public function explodeB() : bool{
+
+				$toIgnite = $this->world->getBlockAt($x, $y, $z);
+
+				if($toIgnite->getTypeId() === VanillaBlocks::AIR()->getTypeId() &&
```
```suggestion
				if($toIgnite->getTypeId() === BlockTypeIds::AIR &&
```

ipad54 commented on src/world/Explosion.php:
```diff
@@ -209,10 +271,26 @@ public function explodeB() : bool{
+
+				if($toIgnite->getTypeId() === VanillaBlocks::AIR()->getTypeId() &&
+					$toIgnite->getSide(Facing::UP)->isSolid()){
+					$this->world->setBlockAt($x, $y, $z, VanillaBlocks::FIRE());
```
You can make a variable with fire block before the loop to save performance

ipad54 commented:
I've mentioned the code paths in the review that aren't related to the respawn anchor. I think you should probably make another PR with your vanilla explosion improvements, because we don't accept multiple changes on different topics in one PR, it's too difficult to review multiple things at once.

b1zeyofficial committed:
Update src/world/Explosion.php

Co-authored-by: ipad54 <63200545+ipad54@users.noreply.github.com>

b1zeyofficial committed:
resolved conversations?

b1zeyofficial committed:
cs

b1zeyofficial commented:
> I've mentioned the code paths in the review that aren't related to the respawn anchor. I think you should probably make another PR with your vanilla explosion improvements, because we don't accept multiple changes on different topics in one PR, it's too difficult to review multiple things at once.

Ok, I'll create two separate PR for this, but for now check for updates regarding the respawn anchor to make things easier for me... Ummm, can I close this?

b1zeyofficial committed:
renamed AnchorChargeSound to RespawnAnchorChargeSound

ipad54 commented:
>  Ummm, can I close this?

Keep this open, just remove your explosion changes that not related to the respawn anchor

ipad54 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -96,6 +96,11 @@ public function getAffectedBlocks() : array{
 	 * @param Block[] $blocks The set of blocks to be affected by fire ignitions
 	 */
 	public function setAffectedBlocks(array $blocks) : void{
+		foreach($blocks as $block){
```
You can just use `Utils::validateArrayValueType()`

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -26,52 +26,50 @@
-
-		if($this->charges >= 1){
+		if($this->charges > self::MIN_CHARGES){
 			$this->explode($player);
```
You need to make `explode()` to accept a nullable parameter.

ipad54 commented on src/world/Explosion.php:
```diff
@@ -106,8 +108,8 @@ public function explodeA() : bool{
-		if($incendiary && !isset($this->fireIgnitions)){
-			$this->fireIgnitions = [];
+		if($incendiary){
+			$this->fireIgnitions = $this->fireIgnitions ?? [];
```
I don’t understand what are you trying to do here, this is not necessary

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
only respawnanchor

b1zeyofficial committed:
PHPStan

b1zeyofficial committed:
cs

b1zeyofficial committed:
Update EntityExplodeEvent.php

b1zeyofficial committed:
property must be private

b1zeyofficial commented:
the only thing i don't like is the explosion resistance of the stone, it shouldn't behave like that, vanilla has more exploding stones

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,92 @@
+use pocketmine\world\sound\RespawnAnchorChargeSound;
+
+final class RespawnAnchor extends Opaque{
+	protected const MIN_CHARGES = 0;
+	protected const MAX_CHARGES = 4;
```
The constants must be private too

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,92 @@
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($item->getTypeId() === ItemTypeIds::fromBlockTypeId(BlockTypeIds::GLOWSTONE) && $this->charges < self::MAX_CHARGES){
+			$this->charges++;
+			$this->position->getWorld()->setBlock($this->position, $this);
```
You can write this in single line by doing `$this->position->getWorld()->setBlock($this->position, $this->setCharges($this->charges + 1));`

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,92 @@
+		return false;
+	}
+
+	public function explode(?Player $player) : void{
```
This function needs to be private

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,92 @@
+	public function explode(?Player $player) : void{
+		$ev = new BlockPreExplodeEvent($this, 5, $player);
+		$ev->setIncendiary(true);
+
```
I don't understand why did you delete the `isCancelled()` check. Bring it back please and just add `$ev->call();` before it

ipad54 commented on src/data/bedrock/block/convert/BlockObjectToStateSerializer.php:
```diff
@@ -1746,6 +1747,9 @@ private function registerSerializers() : void{
 			return Writer::create(Ids::RESIN_CLUMP)
 				->writeFacingFlags($block->getFaces());
 		});
+		$this->map(Blocks::RESPAWN_ANCHOR(), fn(RespawnAnchor $block) => Writer::create(Ids::RESPAWN_ANCHOR)
+		->writeInt(StateNames::RESPAWN_ANCHOR_CHARGE, $block->getCharges())
+		);
```
You should probably make this in a single line or use a full anonymous function

ipad54 commented on src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php:
```diff
@@ -1597,6 +1597,9 @@ private function registerDeserializers() : void{
 		$this->mapStairs(Ids::RESIN_BRICK_STAIRS, fn() => Blocks::RESIN_BRICK_STAIRS());
 		$this->map(Ids::RESIN_BRICK_WALL, fn(Reader $in) => Helper::decodeWall(Blocks::RESIN_BRICK_WALL(), $in));
 		$this->map(Ids::RESIN_CLUMP, fn(Reader $in) => Blocks::RESIN_CLUMP()->setFaces($in->readFacingFlags()));
+		$this->map(Ids::RESPAWN_ANCHOR, fn(Reader $in) => Blocks::RESPAWN_ANCHOR()
+		->setCharges($in->readBoundedInt(StateNames::RESPAWN_ANCHOR_CHARGE, 0, 4))
+		);
```
You should probably make this in a single line or use a full anonymous function

ipad54 commented on src/entity/object/EndCrystal.php:
```diff
@@ -130,6 +130,8 @@ public function explode() : void{
 		$ev->call();
 		if(!$ev->isCancelled()){
 			$explosion = new Explosion($this->getPosition(), $ev->getRadius(), $this);
+			$explosion->setFireChance(0);
```
This isn't necessary. The fire chance is 0 by default

ipad54 commented on src/entity/object/PrimedTNT.php:
```diff
@@ -122,6 +122,8 @@ public function explode() : void{
 		if(!$ev->isCancelled()){
 			//TODO: deal with underwater TNT (underwater TNT treats water as if it has a blast resistance of 0)
 			$explosion = new Explosion(Position::fromObject($this->location->add(0, $this->size->getHeight() / 2, 0), $this->getWorld()), $ev->getRadius(), $this);
+			$explosion->setFireChance(0);
```
Same here

ipad54 commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -0,0 +1,120 @@
+
+	public function __construct(
+		Block $block,
+		protected float $radius,
+		private readonly ?Player $player = null,
+		protected float $fireChance = 0.0
```
All these properties must be private

ipad54 replied:
Also add a fire chance validation here too

ipad54 commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -0,0 +1,120 @@
+	 * Establish the probability of fire
+	 */
+	public function setFireChance(float $fireChance) : void{
+		if($fireChance < 0 || $fireChance > 1){
```
You also need to check that the fire chance isn't NaN or INF

ipad54 commented on src/event/entity/EntityExplodeEvent.php:
```diff
@@ -98,4 +100,27 @@ public function setYield(float $yield) : void{
+	 * @param Block[] $ignitions
+	 */
+	public function setIgnitions(array $ignitions) : void{
+		foreach($ignitions as $ignition){
```
You also need to use `Utils::validateArrayValueType()` here

ipad54 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,147 @@
+	 */
+	public function __construct(
+		Block $block,
+		protected Position $position,
+		protected array $blocks,
+		protected float $yield,
+		private array $ignitions,
+		protected float $fireChance
```
These properties also need to be private

ipad54 replied:
Also add a fire chance validation here too

b1zeyofficial replied:
I made private only `private array $ignitions,
protected float $fireChance` Does it make sense to do this for other properties?

ipad54 replied:
> Does it make sense to do this for other properties?

I don’t see any point why it doesn’t. Private properties can be always turned into protected or public if there is a reason to do that, but not vice versa

ipad54 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,147 @@
+	 * @param Block[] $ignitions The set of blocks to set as ignited
+	 */
+	public function setIgnitions(array $ignitions) : void{
+		foreach($ignitions as $ignition){
```
You also need to use here `Utils::validateArrayValueType()`

ipad54 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,147 @@
+	 * @param float $fireChance The fire chance to set
+	 */
+	public function setFireChance(float $fireChance) : void{
+		if($fireChance < 0.0 || $fireChance > 1.0){
```
Same here about NaN/INF

ipad54 commented on src/world/Explosion.php:
```diff
@@ -150,13 +163,35 @@ public function explodeB() : bool{
-			$ev = new EntityExplodeEvent($this->what, $this->source, $this->affectedBlocks, $yield);
+			$ev = new EntityExplodeEvent($this->what, $this->source, $this->affectedBlocks, $yield, $this->fireIgnitions);
+
+			$ev->setIgnitions($this->fireIgnitions);
```
There is no point to call `$ev->setIgnitions()` because you already passed them in the event constructor

ipad54 commented on src/world/Explosion.php:
```diff
@@ -150,13 +163,35 @@ public function explodeB() : bool{
+				$this->source,
+				$this->affectedBlocks,
+				$yield,
+				$this->affectedBlocks,
```
Why do you use `$this->affectedBlocks` again instead of your fire ignitions array?

b1zeyofficial replied:
After my research, adding `$this->fireIgnitions` gives a load on explosion respawnanchor. Maybe I should remove it and leave only one use of `$this->affectedBlocks`

ipad54 replied:
> After my research, adding `$this->fireIgnitions` gives a load on explosion respawnanchor.

This is because you include air blocks that not affected by the explosion to the potential ignition list. To save performance, you need to add blocks to fire ignition in the place where they are added to the `affectedBlocks`

b1zeyofficial replied:
I don't quite understand where exactly to change the code
```
if($incendiary && isset($this->affectedBlocks[World::blockHash($vBlockX, $vBlockY, $vBlockZ)]) && Utils::getRandomFloat() <= $this->fireChance){
    $this->fireIgnitions[spl_object_id($block)] = $block;
}
```

but I don't think that's true

b1zeyofficial replied:
Check out the latest updates, but the fire has become smaller

ipad54 commented on src/world/Explosion.php:
```diff
@@ -214,11 +249,31 @@ public function explodeB() : bool{
 	}
+
+	public function setFireChance(float $fireChance) : void{
+		$this->fireChance = $fireChance;
```
Validation is missing here

b1zeyofficial committed:
that should solve it

b1zeyofficial committed:
properties must be private. respawn anchor explosion optimization...

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

ipad54 commented on src/entity/object/EndCrystal.php:
```diff
@@ -130,6 +130,7 @@ public function explode() : void{
 		$ev->call();
 		if(!$ev->isCancelled()){
 			$explosion = new Explosion($this->getPosition(), $ev->getRadius(), $this);
+
```
Please remove this line addition

ipad54 commented on src/entity/object/PrimedTNT.php:
```diff
@@ -122,6 +122,7 @@ public function explode() : void{
 		if(!$ev->isCancelled()){
 			//TODO: deal with underwater TNT (underwater TNT treats water as if it has a blast resistance of 0)
 			$explosion = new Explosion(Position::fromObject($this->location->add(0, $this->size->getHeight() / 2, 0), $this->getWorld()), $ev->getRadius(), $this);
+
```
Please remove this line addition

ipad54 commented on src/world/Explosion.php:
```diff
@@ -214,11 +246,34 @@ public function explodeB() : bool{
+				$toIgnite = $this->world->getBlockAt($x, $y, $z);
+
+				if($toIgnite->getTypeId() === BlockTypeIds::AIR &&
+					$toIgnite->getSide(Facing::UP)->isSolid()){
```
Why do you use `UP` here instead of `DOWN`? Also, `isSolid()` isn't fire block support requirement. It requires the supporting block to provide `SupportType::FULL`

b1zeyofficial committed:
delete unnecessary lines and update Explosion.php

b1zeyofficial committed:
oops

b1zeyofficial commented:
I dont understand, is this the expected behavior of fire or not?
![image](https://github.com/user-attachments/assets/8fcef9fa-f63a-46c1-8ae0-c6d069e8b073)

b1zeyofficial committed:
CS

b1zeyofficial committed:
just test

b1zeyofficial committed:
shut up cs

b1zeyofficial committed:
my bad

ipad54 commented on src/world/Explosion.php:
```diff
@@ -150,13 +164,33 @@ public function explodeB() : bool{
+			}
+
+			$yield = $ev->getYield();
+			$this->affectedBlocks = $ev->getBlockList();
```
fire ignitions update from EntityExplodeEvent is missing here

ipad54 commented on src/world/Explosion.php:
```diff
@@ -214,11 +248,34 @@ public function explodeB() : bool{
+
+			$fireBlock = VanillaBlocks::FIRE();
+
+			foreach($this->fireIgnitions as $ignition){
```
I don’t like that you have to iterate over `fireIgnitions` on each `affectedBlocks` iteration. I think you could have `World::blockHash()` as a `fireIgnitions` element key instead of spl_object_id() (similar to how `affectedBlocks` are stored) and then check if `fireIgnitions` has that key on `affectedBlocks` iteration. If so and block below can support fire block, set fire instead of air

ipad54 commented on src/world/Explosion.php:
```diff
@@ -125,6 +136,9 @@ public function explodeA() : bool{
 								if($blastForce > 0){
 									if(!isset($this->affectedBlocks[World::blockHash($vBlockX, $vBlockY, $vBlockZ)])){
 										$_block = $this->world->getBlockAt($vBlockX, $vBlockY, $vBlockZ, true, false);
+										if($incendiary && Utils::getRandomFloat() <= $this->fireChance){
```
This should be done inside the `getAffectedBlocks()` loop to include all block parts if it’s a double block

b1zeyofficial replied:
I'm not sure we need to use World::blockHash twice, instead I would prefer using $affectedHash variable like

ipad54 replied:
Yes, you can create `$posHash` variable

ipad54 commented on src/world/Explosion.php:
```diff
@@ -54,6 +61,9 @@ class Explosion{
 	/** @var Block[] */
 	public array $affectedBlocks = [];
 	public float $stepLen = 0.3;
+	private float $fireChance = 0.0;
```
I would also like to have an ability to pass fire chance in the explosion constructor

b1zeyofficial commented:
> I dont understand, is this the expected behavior of fire or not?

Vanilla 1.21.62
![image](https://github.com/user-attachments/assets/9bf82f20-6077-4176-83d9-527ba7d96699)

b1zeyofficial committed:
Update Explosion.php

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,95 @@
+
+		$this->position->getWorld()->setBlock($this->position, VanillaBlocks::AIR());
+
+		$explosion = new Explosion($this->position, $ev->getRadius(), $this);
```
According to Minecraft Java code, the explosion happens at the centered position, so block X, Y, Z need to be adjusted by 0.5

b1zeyofficial replied:
I don't quite understand this, because we need to create an object `$position = new Position`.. Does it make sense to do this if everything is fine

ipad54 replied:
It's necessary because without adjustment the explosion won't be able to destroy much blocks with good blast resistance (such as stone).

Explosion without adjustment:
![Screenshot_20250316_163905](https://github.com/user-attachments/assets/b319fa77-4cae-4b9a-a828-8038290e098c)

Explosion with adjustment:
![Screenshot_20250316_164045](https://github.com/user-attachments/assets/e1342b50-cffc-4a19-8aa7-a7ae996387cc)

ipad54 commented on src/world/Explosion.php:
```diff
@@ -214,11 +250,35 @@ public function explodeB() : bool{
 				}
 				$this->world->setBlockAt($pos->x, $pos->y, $pos->z, $airBlock);
```
You can just do `$this->world->setBlockAt($pos->x, $pos->y, $pos->z,
				 	isset($this->fireIgnitions[$hash]) &&
					$block->getSide(Facing::DOWN)->getSupportType(Facing::UP) === SupportType::FULL ? $fireBlock : $airBlock
				);` here instead of adding another foreach

ipad54 commented:
The block also needs to be registered in `StringToItemParser`

b1zeyofficial committed:
its all maybe

b1zeyofficial committed:
PHPStan

b1zeyofficial committed:
lol

dktapps commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,147 @@
+	 * @param float $fireChance The fire chance to set
+	 */
+	public function setFireChance(float $fireChance) : void{
+		if(!is_finite($fireChance) || $fireChance < 0.0 || $fireChance > 1.0){
```
Use `Utils::checkFloatNotInfOrNaN()`

dktapps commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,147 @@
+	 *
+	 * @param float $yield The intensity of the explosion (0-100)
+	 */
+	public function setYield(float $yield) : void{
```
This also needs to use `Utils::checkFloatNotInfOrNaN()`

dktapps commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,147 @@
+		}
+		if($fireChance < 0.0 || $fireChance > 1.0){
+			throw new \InvalidArgumentException("Fire chance must be between 0 and 1.");
+		}
```
and again here for yield and fireChance

b1zeyofficial committed:
use checkFloatNotInfOrNaN()

b1zeyofficial committed:
short message

b1zeyofficial committed:
promoteKeys

dktapps commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -53,11 +52,13 @@ public function __construct(
 		}
 		if($fireChance < 0.0 || $fireChance > 1.0){
-			throw new \InvalidArgumentException("Fire chance must be between 0 and 1.");
+			throw new \InvalidArgumentException("Fire chance must be a infinite or nan number between 0 and 1.");
```
You don't ned to mention infinite or nan here, it's already reported by Utils if needed

dktapps commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -139,8 +141,9 @@ public function getFireChance() : float{
-			throw new \InvalidArgumentException("Fire chance must be a finite number between 0 and 1.");
+		Utils::checkFloatNotInfOrNaN("fireChance", $fireChance);
+		if($fireChance < 0.0 || $fireChance > 1.0){
+			throw new \InvalidArgumentException("Fire chance must be a inf or nan number between 0 and 1.");
```
Same again here

dktapps commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -49,8 +49,9 @@ public function __construct(
+		Utils::checkFloatNotInfOrNaN("fireChance", $fireChance);
 		if($fireChance < 0.0 || $fireChance > 1.0){
-			throw new \InvalidArgumentException("Fire chance must be between 0 and 1.");
+			throw new \InvalidArgumentException("Fire chance must be a inf or nan number between 0 and 1.");
```
and here

dktapps commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -109,8 +110,9 @@ public function getFireChance() : float{
-			throw new \InvalidArgumentException("Fire chance must be a finite number between 0 and 1.");
+		Utils::checkFloatNotInfOrNaN("fireChance", $fireChance);
+		if($fireChance < 0.0 || $fireChance > 1.0){
+			throw new \InvalidArgumentException("Fire chance must be a inf or nan number between 0 and 1.");
```
and here

dktapps commented on src/world/Explosion.php:
```diff
@@ -264,8 +264,9 @@ public function explodeB() : bool{
-			throw new \InvalidArgumentException("Fire chance must be a finite number between 0 and 1.");
+		Utils::checkFloatNotInfOrNaN("fireChance", $fireChance);
+		if($fireChance < 0.0 || $fireChance > 1.0){
+			throw new \InvalidArgumentException("Fire chance must be a inf or nan number between 0 and 1.");
```
and here

ipad54 commented on src/world/Explosion.php:
```diff
@@ -199,7 +232,8 @@ public function explodeB() : bool{
 		$airBlock = VanillaBlocks::AIR();
 
-		foreach($this->affectedBlocks as $block){
+		$this->affectedBlocks = Utils::promoteKeys($this->affectedBlocks);
```
You can use promoteKeys() in the foreach instead of this

b1zeyofficial committed:
:)

ipad54 commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -0,0 +1,126 @@
+
+	public function __construct(
+		Block $block,
+		private float $radius,
```
Also check the radius not to be NaN/INF

ipad54 commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -0,0 +1,126 @@
+		return $this->radius;
+	}
+
+	public function setRadius(float $radius) : void{
```
And here

b1zeyofficial committed:
nan/inf check

ipad54 commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -0,0 +1,128 @@
+		if(!$incendiary){
+			$this->fireChance = 0;
+		}elseif($this->fireChance <= 0){
+			$this->fireChance = 1.0 / 3.0;
```
It would be better to create a private `DEFAULT_FIRE_CHANCE` constant with this value

ipad54 commented on src/world/Explosion.php:
```diff
@@ -54,13 +58,16 @@ class Explosion{
 		public float $radius,
-		private Entity|Block|null $what = null
+		private Entity|Block|null $what = null,
+		private float $fireChance = 0.0
```
Also check fire chance here for INF/NaN

ipad54 commented on src/world/Explosion.php:
```diff
@@ -212,13 +245,28 @@ public function explodeB() : bool{
 				if(($t = $this->world->getTileAt($pos->x, $pos->y, $pos->z)) !== null){
 					$t->onBlockDestroyed(); //needed to create drops for inventories
 				}
-				$this->world->setBlockAt($pos->x, $pos->y, $pos->z, $airBlock);
```
You should still set the block inside this condition. Fire can't be placed on the TNT block pos in vanilla

ipad54 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,96 @@
+		}
+
+		if($this->charges > self::MIN_CHARGES){
+			$this->explode($player);
```
Please add `//TODO: Implement the ability to set a respawn point in Nether` before calling explode()

ipad54 commented:
Please remove your changes in composer.lock & fix the tests by running php-cs-fixer

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial commented:
sad

b1zeyofficial committed:
test

b1zeyofficial commented:
I don't understand how to remove the file change, maybe I should just recreate the PR?

b1zeyofficial committed:
cs

b1zeyofficial committed:
one more test

b1zeyofficial committed:
one more test

b1zeyofficial commented:
mm, that's it

b1zeyofficial committed:
forgot to change lol

b1zeyofficial commented:
Is there any point in fixing this PHPStan?

dktapps commented on src/world/Explosion.php:
```diff
@@ -214,11 +250,27 @@ public function explodeB() : bool{
+				$block->getSide(Facing::DOWN)->getSupportType(Facing::UP) === SupportType::FULL
+					? VanillaBlocks::FIRE()
+					: $airBlock
+			);
```
Why are you setting the block twice?

dktapps commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	public function getLightLevel() : int{
+		return $this->charges > 0 ? 3 + 4 * ($this->charges - 1) : 0;
```
Looks like this could just be
```suggestion
		return ($this->charges * 4) - 1;
```

Am I wrong?

ipad54 replied:
You should still check that the charges > 0, otherwise the light level will be -1 for 0 charges

b1zeyofficial replied:
you can take into account the check > 0

b1zeyofficial replied:
i think you can take into account the check > 0

ipad54 replied:
So basically it’ll be `return $this->charges > 0 ? ($this->charges * 4) - 1 : 0;`

dktapps replied:
> You should still check that the charges > 0, otherwise the light level will be -1 for 0 charges

True. Good catch.

b1zeyofficial committed:
optimization?

ipad54 commented on src/event/block/BlockPreExplodeEvent.php:
```diff
@@ -0,0 +1,129 @@
+	use CancellableTrait;
+
+	private bool $blockBreaking = true;
+	private const DEFAULT_FIRE_CHANCE = 1.0 / 3.0;
```
By convention constants are usually placed above properties

ipad54 commented on src/world/Explosion.php:
```diff
@@ -214,11 +251,25 @@ public function explodeB() : bool{
+			if(isset($this->fireIgnitions[$hash]) &&
+				$block->getSide(Facing::DOWN)->getSupportType(Facing::UP) === SupportType::FULL
+			){
+				$this->world->setBlockAt($pos->x, $pos->y, $pos->z, $fireBlock);
```
You are still calling setBlockAt() twice. Do this check in the place where air block is being set (on line 252) and set fire instead if the check is true.

b1zeyofficial committed:
hm

b1zeyofficial committed:
Update RespawnAnchor.php

b1zeyofficial committed:
...

ipad54 committed:
Added return $this

ipad54 committed:
Update documentation

ipad54 committed:
Remove fire chance from BlockExplodeEvent

The explosion impact is calculated at this stage, so modifying this value will do nothing

ipad54 committed:
Document setFireChance() in Explosion

ipad54 committed:
Formatting

ipad54 committed:
Implement the ability to set fire chance in EntityPreExplodeEvent

ipad54 committed:
Update BlockPreExplodeEvent.php

ipad54 committed:
Remove unneeded docs

ipad54 committed:
CS

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

ShockedPlot7560 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,122 @@
+	 * @param Block[] $blocks
+	 */
+	public function setAffectedBlocks(array $blocks) : void{
+		Utils::validateArrayValueType($blocks, fn(Block $block) => null);
```
static fn can be maded probably here ?

dktapps replied:
No need, it's not persisted

ShockedPlot7560 commented on src/event/block/BlockExplodeEvent.php:
```diff
@@ -0,0 +1,122 @@
+	 * @param Block[] $ignitions
+	 */
+	public function setIgnitions(array $ignitions) : void{
+		Utils::validateArrayValueType($ignitions, fn(Block $block) => null);
```
same here ?

dktapps replied:
As above

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

ShockedPlot7560 reviewed (dismissed):
Looks OK to me

dktapps commented on src/world/Explosion.php:
```diff
@@ -198,8 +235,9 @@ public function explodeB() : bool{
+		$fireBlock = VanillaBlocks::FIRE();
 
-		foreach($this->affectedBlocks as $block){
+		foreach(Utils::promoteKeys($this->affectedBlocks) as $hash => $block){
```
This shouldn't be needed, `$hash` is an integer

You probably ought to add `@phpstan-var array<int, Block>` to the property declaration

b1zeyofficial committed:
annotation phpstan var

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

dktapps commented on src/world/Explosion.php:
```diff
@@ -57,7 +57,7 @@ class Explosion{
 	public World $world;
 
-	/** @var Block[] */
+	/** @phpstan-var array<int, Block> */
```
The original should be kept also

b1zeyofficial committed:
Update Explosion.php

dktapps commented on src/world/Explosion.php:
```diff
@@ -57,6 +57,7 @@ class Explosion{
 	private int $rays = 16;
 	public World $world;
 
+	/** @var Block[] */
```
Not like this.
It should be
```
/**
 * @var Block[]
 * @phpstan-var array<int, Block>
 */
```

b1zeyofficial replied:
oops, I updated the file from the browser editor, my bad

b1zeyofficial committed:
Update Explosion.php

b1zeyofficial committed:
CS

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

dktapps commented:
Per internal discussion we think it would be a good idea to have a PlayerRespawnAnchorUseEvent with actions EXPLODE, SET_SPAWN and NONE. The default behaviour in the absence of dimensions would be EXPLODE, but this would allow plugins to alter the behaviour based on custom conditions (and perhaps custom dimensions).

b1zeyofficial commented:
> Per internal discussion we think it would be a good idea to have a PlayerRespawnAnchorUseEvent with actions EXPLODE, SET_SPAWN and NONE. The default behaviour in the absence of dimensions would be EXPLODE, but this would allow plugins to alter the behaviour based on custom conditions (and perhaps custom dimensions).

Maybe with this I can implement respawn set point for respawn achor? In nether
But nether is not implemented yet, I will use DimensionsIds. How do you look at this?

dktapps commented:
No, I'm saying that the logic for setting the spawn point should be implemented, but that the behaviour should be to EXPLODE by default. This will let plugins change it without having to implement extra logic themselves.

b1zeyofficial committed:
added PlayerRespawnAnchorUseEvent and sounds that are needed for its vanilla?

b1zeyofficial committed:
CS

b1zeyofficial commented:
I found a problem that I can't solve. When exploding, using glowstone, when we charge the anchor for 5 charges, it explodes but the GLOWSTONE block itself is placed, this is not Vanilla behavior. I dont understand what the problem is.. 

P.S: This works on all blocks that can be placed

dktapps commented on src/block/RespawnAnchor.php:
```diff
@@ -68,9 +73,45 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+
+			$ev = new PlayerRespawnAnchorUseEvent($player, $this, PlayerRespawnAnchorUseEvent::ACTION_EXPLODE);
+
+			if($this->position->getWorld()->getId() === DimensionIds::NETHER){
```
I said _don't_ do this. This is never going to be the case.

The aim is to make plugins able to change the outcome.

nicholass003 replied:
Also ```World->getId()``` doesn't return ```DimensionIds```, it returns a unique world identifier.
https://github.com/pmmp/PocketMine-MP/blob/e88b81a4cbd9274aa3b6faa0b3f847c13ca0ecd5/src/world/World.php#L616

dktapps replied:
Honestly we probably ought to deprecate this function. It serves no obvious purpose. But that's a topic for another issue.

dktapps commented on src/block/RespawnAnchor.php:
```diff
@@ -68,9 +73,45 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+
+						$player->setSpawn($this->position);
+						$this->position->getWorld()->addSound($this->position, new RespawnAnchorSetSpawnSound());
+						$player->sendMessage(TextFormat::GRAY . "Respawn point set");
```
This needs translations

b1zeyofficial replied:
can i create PR Language for this now?

dktapps commented on src/block/RespawnAnchor.php:
```diff
@@ -68,9 +73,45 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+					return false;
+
+				case PlayerRespawnAnchorUseEvent::ACTION_SET_SPAWN:
+					if($this->charges >= 4){
```
This should use a constant

nicholass003 commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,139 @@
+					return false;
+
+				case PlayerRespawnAnchorUseEvent::ACTION_SET_SPAWN:
+					if($this->charges >= 4){
```
To set the player spawn point, it doesn't have to be full, use this
 ```php
if($this->charges > self::MIN_CHARGES)
```

b1zeyofficial replied:
We still need to use >= because the logic requires the number of charges to be equal to or greater than the max value to set the respawn point. Otherwise the respawn point may not be set

nicholass003 replied:
> We still need to use >= because the logic requires the number of charges to be equal to or greater than the max value to set the respawn point. Otherwise the respawn point may not be set

Where did you get this logic from?
This is the basis that I took as a reference.
https://minecraft.wiki/w/Respawn_Anchor#Respawning

b1zeyofficial replied:
My bad, i thought that the anchor should be fully charged before setting the respawn point. I checked the videos on YouTube how it works in gameplay

b1zeyofficial committed:
Update RespawnAnchor.php

b1zeyofficial committed:
instead of min we make max

b1zeyofficial commented:
im thinking of making it so that before each player dies, the anchor charge is taken away and the sound RespawnAnchorDepleteSound is played

b1zeyofficial committed:
we still use >=

b1zeyofficial committed:
oops

b1zeyofficial committed:
lmao

b1zeyofficial committed:
...

b1zeyofficial committed:
wtf

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

dktapps commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,126 @@
+
+			$ev = new PlayerRespawnAnchorUseEvent($player, $this, PlayerRespawnAnchorUseEvent::ACTION_EXPLODE);
+			$ev->call();
+			if($ev->isCancelled()){
```
Having this as well as NONE doesn't make any sense.

dktapps commented on src/block/RespawnAnchor.php:
```diff
@@ -0,0 +1,126 @@
+
+					$player->setSpawn($this->position);
+					$this->position->getWorld()->addSound($this->position, new RespawnAnchorSetSpawnSound());
+					$player->sendMessage(TextFormat::GRAY . "Respawn point set");
```
Still need translations for this

dktapps reviewed (changes requested):
It doesn't make any sense for the event to be both cancellable and have a NONE action

b1zeyofficial committed:
added translations for RespawnAnchor
removed ACTION_NONE

b1zeyofficial committed:
...

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

dktapps reviewed (dismissed):
Looks good, tested and seems working well

dktapps committed:
Merge branch 'minor-next' into respawn-anchor

ShockedPlot7560 committed:
Merge branch 'minor-next' into respawn-anchor

b1zeyofficial committed:
expected 6000

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor

dktapps committed:
Merge branch 'minor-next' into respawn-anchor

Yexeed commented:
This pr actually breaks compatibility, which was never specified, neither here or changelog: EntityExplodeEvent got new (required) field in it's constructor, breaking plugins that was using it earlier.

Yexeed commented:
Also, Explosion.php got new field, which is private. How did it even get through? It literally says good luck with expanding and making custom logic for explosions (e.g. different calculation, overriding functions etc.)
And another: BlockExplodeEvent (which is new, that's ok), seems to have similar fields to EntityExplodeEvent: `$blocks` and `$yield`. However, those were `protected` in EntityExplode, but in BlockExplode, they are `private`. Is this also ok, @dktapps ?

dktapps commented:
Private is favoured unless there's a specific reason to increase their visibility. We can increase visibility from `private` without breaking compatibility, but the reverse breaks BC. In the case of events there's not really a good reason to make them more than `private` anyway.

In the case of `Explosion`, it's a mess anyway. It already has private fields so this isn't an isolated case. I wouldn't be against increasing the visibility though.

Good catch on the `EntityExplodeEvent` BC break though. I guess no one spotted that.
